cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,1060 @@
+contains the APIs for functions related to copying from one column to another. Note the  `.hpp` 
+file extension used to indicate a C++ header file.
+
+Header files should use the `#pragma once` include guard. 
+
+The naming of external API headers should be consistent with the name of the folder that contains 
+the source files that implement the API. For example, the implementation of the APIs found in
+`cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 
+the APIs reside in `cudf/tests/copying/`.
+
+Internal API headers containing `detail` namespace definitions that are used across translation 
+units inside libcudf should be placed in `include/cudf/detail`.
+
+## File extensions
+
+- `.hpp` : C++ header files
+- `.cpp` : C++ source files
+- `.cu`  : CUDA C++ source files
+- `.cuh` : Headers containing CUDA device code
+
+Only use `.cu` and `.cuh` if necessary. A good indicator is the inclusion of `__device__` and other
+symbols that are only recognized by `nvcc`. Another indicator is Thrust algorithm APIs with a device
+execution policy (always `rmm::exec_policy` in libcudf).
+
+## Code and Documentation Style and Formatting
+
+libcudf code uses [snake_case](https://en.wikipedia.org/wiki/Snake_case) for all names except in a 
+few cases: template parameters, unit tests and test case names may use Pascal case, aka 
+[UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case). We do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), except sometimes when naming device data variables and their corresponding
+host copies. Private member variables are typically prefixed with an underscore.
+
+```c++
+template <typename IteratorType>
+void algorithm_function(int x, rmm::cuda_stream_view s, rmm::device_memory_resource* mr) 
+{
+  ...
+}
+
+class utility_class 
+{
+  ...
+ private:
+  int _rating{};
+  std::unique_ptr<cudf::column> _column{};
+}
+
+TYPED_TEST_CASE(RepeatTypedTestFixture, cudf::test::FixedWidthTypes);
+
+TYPED_TEST(RepeatTypedTestFixture, RepeatScalarCount)
+{
+  ...
+}
+```
+
+C++ formatting is enforced using `clang-format`. You should configure `clang-format` on your 
+machine to use the `cudf/cpp/.clang-format` configuration file, and run `clang-format` on all 
+changed code before committing it. The easiest way to do this is to configure your editor to 
+"format on save".
+
+Aspects of code style not discussed in this document and not automatically enforceable are typically
+caught during code review, or not enforced.
+
+### C++ Guidelines
+
+In general, we recommend following 
+[C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). We also 
+recommend watching Sean Parent's [C++ Seasoning talk](https://www.youtube.com/watch?v=W2tWOdzgXHA), 
+and we try to follow his rules: "No raw loops. No raw pointers. No raw synchronization primitives." 
```

Comment:
AFAIK the pointer rule only refers to owning pointers. Passing pointers around without changing the ownership should be fine.
```suggestion
and we try to follow his rules: "No raw loops. No raw owning pointers. No raw synchronization primitives." 
```

Reply:
But it's a direct quote. And I think my bullet point after the quote clears this up. In general I don't think I agree. You will notice that we don't have raw pointers at all in our APIs unless necessary - this is what our view classes help with.

Reply:
Did not realize it's literally a quote, disregard the comment then. Just trying to make the guideline clearer. 
Further down in the document we do recommend to pass in/out parameters as pointers (unless it's a type with a view).